### PR TITLE
[WIP] ipa-join: allowing call with jsonrpc into freeipa API

### PR DESCRIPTION
**Pull request by mulatinho**
- Adding JSON-C and LibCURL library into configure.ac and Makefile.am
- Creating a API call with option '-j' or '--jsonrpc' to make host join on FreeIPA with JSONRPC and libCURL.

TODO: unenroll process with JSONRPC.

To test the call:
```
# kinit admin
# ipa-join -s server.freeipa.ipadomain -j
```

Debug:
```
# ipa-join -s server.freeipa.ipadomain -j -d
```
Results: https://pastebin.com/0h3eQXHV

Related: https://pagure.io/freeipa/issue/7966

### CODE REVIEW

**Comment by abbra** [comment]
This looks promising, thank you, @mulatinho. Few comments.

* please add build dependencies for json-c into freeipa.spec.in
* I think we should not introduce the additional command line switches. Instead, we should be trying json-rpc first, then xml-rpc if the latter failed.
* error handling is missing in many places.
* Eventually we should remove xml-rpc support. I checked that ipa-3-0 already supports json-rpc calls on the server side.

**Comment by tiran** [comment]
Good start!

* Please don't set any TLS/SSL related option. The current patch disables cert validation and sets TLS 1.2. We want to enforce cert checks and use the system policy for TLS/SSL related options.
* To simplify error checking I suggest to use macros and ``goto`` with a cleanup block. You could e.g. define a macro ``CURL_SETOPT(curl, opt, val)`` that performs setting, error checking, error reporting, and goto cleanup.
* print debug output to ``stderr``

**Comment by mulatinho** [comment]
thanks for the feedback @abbra and @tiran. christian, I put TLSv1.2 fixed there mainly because it was on the xmlrpc version, but this is not a problem, I will avoid the code part which skip cert checks and the TLSv1.2 version, thanks for the tip of cleanup, I will try for sure :)

my life is not so easy now, my wife is on late pregnancy and I have a kid of two years old that loves to play :D (thanks God for all that). But I will finish this today or tomorrow and I will make the commit rebase with the changes you asked.

**Comment by mulatinho** [comment]
Hello! Back again :D So to fix this issue the best is to try call with JSON RPC first and if we got an error try XML RPC? 

> Eventually we should remove xml-rpc support. I checked that ipa-3-0 already supports json-rpc calls on the server side.

Or do I need to replace the XMLRPC content?

**Comment by abbra** [comment]
> Hello! Back again :D So to fix this issue the best is to try call with JSON RPC first and if we got an error try XML RPC?
> 
> > Eventually we should remove xml-rpc support. I checked that ipa-3-0 already supports json-rpc calls on the server side.
> 
> Or do I need to replace the XMLRPC content?

My thinking was to enable both (JSON RPC first to call, then XMLRPC if failed). However, make it a configure option and at some point later we can drop the option from build. And even further in future we remove XMLRPC support completely.

**Comment by abbra** [comment]
I still like to get this PR merged. @mulatinho, I can help with this, if do not mind.

**Comment by mulatinho** [comment]
Hey guys! Sorry for the late reply (classic), how are you doing all? I
really hope this get merged someday, I am a fan of this project and would
love to be part of it with more contributions, I will try take a look on
this, like I said before, these days I am little busy with my kids and now
a new job :)


To be honest I was almost forgot of this PR :P I would love to hear some
ideas from you all, this is working but I need to implement some checks,
and finish one other part.

**Comment by tiran** [comment]
Hey,

in case you are still interested to work on the patch, then I suggest to replace the JSON parser with jansson. Its high level ``json_pack`` and ``json_unpack`` methods are such much nicer and easier to use than libjson:

```
json_t *root;
json_error_t error;
int res;
const char *host, *princ;
size_t *hostlen, *princlen;

root = json_loads(s, 0, &error);
if (root == NULL) { ... }
int = json_unpack_ex(root, &error, JSON_STRICT, "[s%, s%]", &host, &hostlen, &princ, &princlen);
if (res == -1) { ... }

strndup(host, hostlen);
strndup(princ, princlen);

json_decref(root);
```

File: configure.ac [hunk 1]
```diff
@@ -178,6 +178,42 @@ if test x"$try_xmlrpc_fallback" = xtrue; then
     AC_SUBST(XMLRPC_LIBS)
 fi
 
+dnl ---------------------------------------------------------------------------
+dnl - Check for LIBCURL
+dnl ---------------------------------------------------------------------------
+PKG_CHECK_MODULES([LIBCURL], [libcurl], [],
```
Comment (tiran) [inline configure.ac:184]: You can keep it simple and assume that platforms have proper pkg-config support for necessary libraries:

```
dnl ---------------------------------------------------------------------------
dnl - Check for jansson and libcurl for ipa-join
dnl ---------------------------------------------------------------------------
PKG_CHECK_MODULES([JANSSON], [jansson])
PKG_CHECK_MODULES([LIBCURL], [libcurl])
```

**Comment by Carbenium** [comment]
I gave it a shot and tried to implement the comments from @abbra and @tiran.
See https://github.com/freeipa/freeipa/compare/master...Carbenium:ipa-join-json-rpc

If this is acceptable so far and @mulatinho agrees, I could try to implement the unenrollment as well.

**Comment by abbra** [comment]
@Carbenium sounds good. Please submit as a pull request so that we can discuss the proposed variant on its own.

**Comment by tiran** [comment]
Thank you for your pull request. https://github.com/freeipa/freeipa/pull/4724 now has landed in master. I'm closing this PR as duplicate.

**Pull request closed without merging**